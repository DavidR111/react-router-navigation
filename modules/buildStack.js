/* @flow */
/* eslint no-param-reassign: 0*/

import { Children, cloneElement } from 'react'
import { Platform } from 'react-native'
import type { CardProps, TabProps } from './TypeDefinitions'

export default function buildStack<Item: CardProps | TabProps>(
  children: Array<React$Element<Item>>
): Array<Item & { key: string }> {
  return Children.toArray(children).reduce((stack, child) => {
    const item = Object.keys(child.props).reduce((props, key) => {
      if (key === 'path') {
        props.key = child.props[key]
        props[key] = child.props[key]
      } else if (key === 'render' || key === 'component' || key === 'children') {
        const renderMethod = (ownProps: any) => cloneElement(child, ownProps)
        // @TODO (hot fix: rename key for jest test)
        if (Platform.OS === 'ios') Object.defineProperty(renderMethod, 'name', { value: key })
        props[key] = renderMethod
      } else {
        props[key] = child.props[key]
      }
      return props
    }, {})
    if (!item.path) return stack
    return [...stack, item]
  }, [])
}
