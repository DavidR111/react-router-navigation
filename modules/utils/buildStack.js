/* @flow */
/* eslint no-param-reassign: 0*/

import { Children, cloneElement } from 'react'

export default function buildStack<Item>(
  children: Array<React$Element<Item>>
): Array<Item & { key: string }> {
  return Children.toArray(children).reduce((stack, child) => {
    const item = Object.keys(child.props).reduce((props, key) => {
      if (key === 'path') {
        props.key = child.props[key]
        props[key] = child.props[key]
      } else if (key === 'render' || key === 'component' || key === 'children') {
        const renderMethod = (ownProps: any) => cloneElement(child, ownProps)
        Object.defineProperty(renderMethod, 'name', { value: key })
        props[key] = renderMethod
      } else {
        props[key] = child.props[key]
      }
      return props
    }, {})
    if (!item.path) return stack
    return [...stack, item]
  }, [])
}
