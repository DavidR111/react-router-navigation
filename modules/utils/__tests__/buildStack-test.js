import React, { createElement, cloneElement } from 'react'
import { Text } from 'react-native'
import { Router, Route } from 'react-router'
import createHistory from 'history/createMemoryHistory'
import renderer from 'react-test-renderer'
import buildStack from './../buildStack'

describe('buildStack() util', () => {
  it('buildStack() works correctly', () => {
    const children = [
      <Route
        path="hey"
        render={() => null}
      />,
      <Route
        path="hello"
        component={() => null}
        title="Hello"
      />
    ]
    const results = [{
      key: 'hey',
      path: 'hey',
      render: () => nulll,
    }, {
      key: 'hello',
      path: 'hello',
      component: () => null,
      title: 'Hello',
    }]
    expect(JSON.stringify(buildStack(children))).toEqual(JSON.stringify(results))
  })

  it('buildStack() creates correctly render method', () => {
    const children = [
      <Route
        path="/"
        render={({ path }) => <Text>{path}</Text>}
      />
    ]
    const stack = buildStack(children)
    const history = createHistory()
    const component = renderer.create(
      <Router history={history}>
        {stack[0].render()}
      </Router>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
