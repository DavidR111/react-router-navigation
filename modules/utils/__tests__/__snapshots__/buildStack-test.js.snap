exports[`buildStack() util buildStack() creates correctly render method 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail">
  /
</Text>
`;
